rfd/0049-desktop-clipboard.md

```diff
@@ -0,0 +1,197 @@
+          |            Format Data Request PDU |
+          |                 (request for data) |
+          |<-----------------------------------|
+          |                                    |
+          | Format Data Response PDU           |
+          | (clipboard data returned)          |
+          |----------------------------------->|
+          |                                    |
+```
+
+This approach differs from protocols like
+[Remote Frame Buffer](https://datatracker.ietf.org/doc/html/rfc6143) (VNC) or
+[Teleport Desktop Protocol](./0037-desktop-access-protocol.md), where the
+clipboard data is sent along with the notification that there was an update.
+
+### Security
+
+The sharing of a clipboard between two machines requires a high level of trust.
+An untrusted peer or misbehaving RDP server can:
+
+- monitor all clipboard activity on the other end of the connection by
+  requesting a "paste" any time it receives a notification of a clipboard update
+- replace or alter clipboard data before sending it to the remote machine to be
+  pasted
+
+These are _features_ of RDP - clipboard sharing wouldn't work without them,
+though it highlights the importance of trusting the other end of the connection.
+
+For these reasons, clipboard access will be enabled via the `desktop_clipboard`
+role option as specified in [RFD 33](./0033-desktop-access.md). This option will
+default to `true`, but the presence of a single role that disables the clipboard
+will turn it off. When clipboard access is not enabled, Teleport will not
+respond to any clipboard messages received by either the browser or the RDP
+server.
+
+### User Experience
+
+The desktop access client runs in a web browser, which provides limited access
+to the system clipboard. The
+[asynchronous clipboard API](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API)
+provides read and write access to the system clipboard, and is an obvious
+candidate for Teleport. Unfortunately, this API is relatively new and browser
+support varies widely. Access to this clipboard API is meant to be protected by
+the
+[permissions API](https://developer.mozilla.org/en-US/docs/Web/API/Permissions_API).
+At the time of this writing:
+
+- The Firefox permissions API does not support the `clipboard-read` or
+  `clipboard-write` permissions.
+- Safari does not support the permissions API at all, and instead implements
+  [its own restrictions](https://developer.mozilla.org/en-US/docs/Web/API/Permissions_API)
+  on the clipboard API. Most importantly, the system clipboard can only be read
+  from or written to in response to a user gesture.
+
+Until browser support is improved, using the clipboard API with these
+restrcitions would require that we maintain a large amount of browser-specific
+code. Additionally, even with browser-specific code, some browsers simpy won't
+work, as we need:
+
+- access to the clipboard upon receiving a TDP message, which will be decoupled
```

Comment:
I don't understand what this bullet point means

Reply:
In some browsers, we can't use clipboard API to write to the clipboard when the browser receives a "clipboard data" TDP message. This is because the browser only allows access to the clipboard in the context of a click or touch event handler.

Reply:
I see, my fault I just wasn't connecting it to the mention of user gestures above.

Regarding this, the MDN documentation is a bit unclear. For example, we're told of the [Clipboard API](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard)

> Calls to the methods of the Clipboard object will not succeed if the user hasn't granted the needed permissions using the Permissions API and the "clipboard-read" or "clipboard-write" permission as appropriate.

If we look at the [Permissions API Browser compatibility table](https://developer.mozilla.org/en-US/docs/Web/API/Permissions_API#browser_compatibility), it looks like "clipboard-read" and "clipboard-write" are hard no's for Firefox and Safari. However, if we look at the [Clipboard's Browser compatibility table](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API#clipboard), it looks like `readText` and `writeText` are fully supported in all browsers except Firefox. I'm not sure how to square that circle with Safari.

Additionally, I only saw a note on Firefox that it's extensions require user-initiated event callbacks for `writeText`'s, however above you indicate that this is also the case for Safari.